File: src-tauri/src/commands/extension.rs
use tauri::{command, State};

use crate::{
    contexts::extension::{
        CustomCheck, CustomCheckParams, CustomExtractor, CustomExtractorParams,
    },
    contexts::tags::{Tag, TagRegistry, TagScope},
    error::CommandError,
    lifecycle::app_state::AppState,
};

#[command]
#[specta::specta]
pub async fn list_custom_checks(
    app_state: State<'_, AppState>,
) -> Result<Vec<CustomCheck>, CommandError> {
    app_state
        .extension_repo
        .list_checks()
        .await
        .map_err(CommandError::from)
}

#[command]
#[specta::specta]
pub async fn create_custom_check(
    params: CustomCheckParams,
    app_state: State<'_, AppState>,
) -> Result<CustomCheck, CommandError> {
    app_state
        .extension_repo
        .create_check(&params)
        .await
        .map_err(CommandError::from)
}

#[command]
#[specta::specta]
pub async fn update_custom_check(
    id: String,
    params: CustomCheckParams,
    app_state: State<'_, AppState>,
) -> Result<CustomCheck, CommandError> {
    app_state
        .extension_repo
        .update_check(&id, &params)
        .await
        .map_err(CommandError::from)
}

#[command]
#[specta::specta]
pub async fn delete_custom_check(
    id: String,
    app_state: State<'_, AppState>,
) -> Result<(), CommandError> {
    app_state
        .extension_repo
        .delete_check(&id)
        .await
        .map_err(CommandError::from)
}

// --- Tags ---

/// Return the full tag catalog so the frontend can render tag pickers
/// / autocomplete in the custom-check editor, template editor, and the
/// Settings → Tags panel.
///
/// `scope` is optional: when present the result is filtered to tags
/// valid in that authoring surface (e.g. `CheckField`). When absent
/// every tag is returned.
#[command]
#[specta::specta]
pub async fn list_tags(
    scope: Option<TagScope>,
    app_state: State<'_, AppState>,
) -> Result<Vec<Tag>, CommandError> {
    let registry = TagRegistry::build(app_state.extension_repo.as_ref())
        .await
        .map_err(CommandError::from)?;
    match scope {
        Some(s) => Ok(registry.in_scope(s).into_iter().cloned().collect()),
        None => Ok(registry.into_tags()),
    }
}

// --- Custom Extractors ---

#[command]
#[specta::specta]
pub async fn list_custom_extractors(
    app_state: State<'_, AppState>,
) -> Result<Vec<CustomExtractor>, CommandError> {
    app_state
        .extension_repo
        .list_extractors()
        .await
        .map_err(CommandError::from)
}

#[command]
#[specta::specta]
pub async fn create_custom_extractor(
    params: CustomExtractorParams,
    app_state: State<'_, AppState>,
) -> Result<CustomExtractor, CommandError> {
    app_state
        .extension_repo
        .create_extractor(&params)
        .await
        .map_err(CommandError::from)
}

#[command]
#[specta::specta]
pub async fn update_custom_extractor(
    id: String,
    params: CustomExtractorParams,
    app_state: State<'_, AppState>,
) -> Result<CustomExtractor, CommandError> {
    app_state
        .extension_repo
        .update_extractor(&id, &params)
        .await
        .map_err(CommandError::from)
}

#[command]
#[specta::specta]
pub async fn delete_custom_extractor(
    id: String,
    app_state: State<'_, AppState>,
) -> Result<(), CommandError> {
    app_state
        .extension_repo
        .delete_extractor(&id)
        .await
        .map_err(CommandError::from)
}

File: src/api/extension.ts
import { commands } from "@/src/bindings";
import type {
  CustomCheck,
  CustomCheckParams,
  CustomExtractor,
  CustomExtractorParams,
  Tag,
  TagScope,
} from "@/src/bindings";

export type { CustomCheck, CustomCheckParams, CustomExtractor, CustomExtractorParams, Tag, TagScope };

// --- Tags ---

/** Fetch the full tag catalog, optionally filtered by scope. */
export async function listTags(scope?: TagScope): Promise<Tag[]> {
  const res = await commands.listTags(scope ?? null);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to list tags");
}

// --- Custom Checks ---

export async function listCustomChecks(): Promise<CustomCheck[]> {
  const res = await commands.listCustomChecks();
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to list custom checks");
}

export async function createCustomCheck(params: CustomCheckParams): Promise<CustomCheck> {
  const res = await commands.createCustomCheck(params);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to create custom check");
}

export async function updateCustomCheck(id: string, params: CustomCheckParams): Promise<CustomCheck> {
  const res = await commands.updateCustomCheck(id, params);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to update custom check");
}

export async function deleteCustomCheck(id: string): Promise<void> {
  const res = await commands.deleteCustomCheck(id);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to delete custom check");
}

// --- Custom Extractors ---

export async function listCustomExtractors(): Promise<CustomExtractor[]> {
  const res = await commands.listCustomExtractors();
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to list custom extractors");
}

export async function createCustomExtractor(params: CustomExtractorParams): Promise<CustomExtractor> {
  const res = await commands.createCustomExtractor(params);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to create custom extractor");
}

export async function updateCustomExtractor(id: string, params: CustomExtractorParams): Promise<CustomExtractor> {
  const res = await commands.updateCustomExtractor(id, params);
  if (res.status === "ok") return res.data;
  throw new Error(res.error ?? "Failed to update custom extractor");
}

export async function deleteCustomExtractor(id: string): Promise<void> {
  const res = await commands.deleteCustomExtractor(id);
  if (res.status !== "ok") throw new Error(res.error ?? "Failed to delete custom extractor");
}
